### PR TITLE
Globalise `CountDifficultStrains` within StrainSkill

### PR DESCRIPTION
**Pull request by Lawtrohux**
Refactor the `CountDifficultStrains` method to be accessible and usable by all rulesets;

- Moves `CountDifficultStrains` into StrainSkill
- Moves storing individual strains into StrainSkill

This globalisation is intended primarily for use within osu!taiko future changes but also to give other rulesets the opportunity to utilise the concept in their own calculations.

While the weighting constants were designed for osu! to give nice values, values look fine - specifically osu!taiko.

### CODE REVIEW

File: osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs [hunk 1]
```diff
@@ -57,7 +57,25 @@ public sealed override void Process(DifficultyHitObject current)
...
+
+            // Store the strain value for the object
+            ObjectStrains.Add(strain);
+        }
+
+        /// <summary>
+        /// Calculates the number of strains weighted against the top strain.
+        /// The result is scaled by clock rate as it affects the total number of strains.
+        /// </summary>
+        public double CountDifficultStrains()
```
Comment (tsunyoku) [inline osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs:71]: we may want to make this virtual in the future to allow ruleset calculators to define their own summation if they want, although it seems the concept works fine enough outside of osu! ruleset

Comment (tsunyoku) [inline osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs:71]:
```suggestion
        public virtual double CountDifficultStrains()
```

following earlier comment

**Review by tsunyoku — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs [hunk 2]
```diff
@@ -26,10 +26,10 @@ public abstract class StrainSkill : Skill
         protected virtual int SectionLength => 400;
 
         private double currentSectionPeak; // We also keep track of the peak strain level in the current section.
-
         private double currentSectionEnd;
 
         private readonly List<double> strainPeaks = new List<double>();
+        protected List<double> ObjectStrains = new List<double>(); // Store individual strains
```
[the remark this answers is not in the document]
Comment (tsunyoku) [reply osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs:32]: not so sure about that, if we're allowing `CountDifficultStrains` to be virtual then any inheritors will need to be able to access the list

[the remark this answers is not in the document]
Comment (Lawtrohux) [reply osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs:32]: is used by `relevantNoteCount` in speed

File: osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs (hunk 1 above)
Comment (tsunyoku) [reply osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs:71]: I'd agree, we should probably rename any existing difficulty attributes if we're going to rename the method though

Comment (stanriders) [reply osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs:71]: AFAIK it's not an issue? We had attribute renames before so it should be fine

Comment (Lawtrohux) [reply osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs]: Yeah can't think of a better one than `CountTopWeightedStrains`, so I'll go with it

**Review by tsunyoku — APPROVED** [review]

**Comment by smoogipoo** [comment]
!diffcalc

**Comment by smoogipoo** [comment]
!diffcalc

**Comment by smoogipoo** [comment]
!diffcalc

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**